common/cpp/src/google_smart_card_common/messaging/typed_message_router.cc

```diff
@@ -43,15 +48,36 @@ void TypedMessageRouter::RemoveRoute(TypedMessageListener* listener) {
   route_map_.erase(route_map_iter);
 }
 
-bool TypedMessageRouter::OnMessageReceived(const pp::Var& message) {
-  std::string error_message;
+bool TypedMessageRouter::OnMessageReceived(Value message,
+                                           std::string* error_message) {
+  // TODO: Remove this conversion and use `Value` directly.
+  const pp::Var pp_message = ConvertValueToPpVar(message);
+
   TypedMessage typed_message;
-  if (!VarAs(message, &typed_message, &error_message)) return false;
+  // TODO: Delete `local_error_message` in favor of (optional) `error_message`.
+  std::string local_error_message;
+  if (!VarAs(pp_message, &typed_message, &local_error_message)) {
+    if (error_message) {
+      *error_message = FormatPrintfTemplate("Cannot parse typed message: %s",
+                                            local_error_message.c_str());
+    }
```

Comment:
nit: this is not sepcific to this CL. These 3 lines are repeated in quite some places, and is kinda interfering with the main logic.

As a suggestion (feel free to discard): Can we encapsulate the if inside the `FormatPrintfTemplate` itself and pass the error_message parameter along: it would be something like: `FormatPrintfTemplate(char* error_message, const char* format, ...)`

Reply:
That's a good point.

The downside will be that the formatting arguments will always be calculated, but, practically, the performance penalty should be negligible and only affecting "unsuccessful" flows. Also there are some tricky cases like `FormatPrintfTemplate(error_message, ..., error_message->c_str())` that would crash if we dereference the null, but those can be solved by putting that nested error into a separate local variable.

So overall I agree that helper is what we should do; will prepare a separate PR soon. Thanks.